src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -53798,11 +53771,315 @@ static void F(A a, B b)
+        var u2 = ((object x, (B, S?) y))t2; // 3
+        u2.y.Item1.ToString(); // 4
+        u2.y.Item2.Value.F.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            // https://github.com/dotnet/roslyn/issues/32599: Track state across tuple element conversions.
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        [WorkItem(32599, "https://github.com/dotnet/roslyn/issues/32599")]
+        [WorkItem(32600, "https://github.com/dotnet/roslyn/issues/32600")]
+        public void Conversions_TupleConversions_13()
+        {
+            var source =
+@"class A
+{
+    public static implicit operator B(A a) => throw null!;
+}
+class B
+{
+}
+struct S
+{
+    internal object? F;
+    internal object G;
+}
+class Program
+{
+    static void F()
+    {
+        A x = new A();
+        S y = new S() { F = 1, G = null }; // 1
+        var t = (x, y, x, y, x, y, x, y, x, y);
+        (B?, S?, B?, S?, B?, S?, B?, S?, B?, S?) u = t;
+        u.Item1.ToString();
+        u.Item2.Value.F.ToString();
+        u.Item2.Value.G.ToString(); // 2
+        u.Item9.ToString();
+        u.Item10.Value.F.ToString();
+        u.Item10.Value.G.ToString(); // 3
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (18,36): warning CS8625: Cannot convert null literal to non-nullable reference type.
+                //         S y = new S() { F = 1, G = null }; // 1
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(18, 36),
+                // (23,9): warning CS8602: Possible dereference of a null reference.
+                //         u.Item2.Value.G.ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "u.Item2.Value.G").WithLocation(23, 9),
+                // (26,9): warning CS8602: Possible dereference of a null reference.
+                //         u.Item10.Value.G.ToString(); // 3
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "u.Item10.Value.G").WithLocation(26, 9));
+        }
+
+        [Fact]
+        [WorkItem(32599, "https://github.com/dotnet/roslyn/issues/32599")]
+        [WorkItem(32600, "https://github.com/dotnet/roslyn/issues/32600")]
+        public void Conversions_TupleConversions_14()
+        {
+            var source =
+@"class A
+{
+    public static implicit operator A?(int i) => new A();
+}
+class B
+{
+    public static implicit operator B(int i) => new B();
+}
+class Program
+{
+    static void F((int, int) t)
+    {
+        (A, B?) u = t; // 1
+        u.Item1.ToString(); // 2
+        u.Item2.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            // https://github.com/dotnet/roslyn/issues/32599: Track state across tuple element conversions.
+            comp.VerifyDiagnostics();
         }
```

Comment:
>} [](start = 8, length = 1)

not related to this PR: it may be good to add a test showing the user-defined conversions on tuples are ignored for analysis (we don't expect users to define their own ValueTuple types)

Reply:
I'm including tests for user-defined conversions between tuples and custom types in the next PR.

---
In reply to: [268850928](https://github.com/dotnet/roslyn/pull/34303#discussion_r268850928) [](ancestors = 268850928)